python/tests/io/test_pyarrow.py

```diff
@@ -0,0 +1,126 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os
+import tempfile
+from unittest.mock import MagicMock
+
+import pytest
+from pyarrow.fs import FileType
+
+from iceberg.io.pyarrow import PyArrowFile
+
+
+def test_pyarrow_input_file():
+    """Test reading a file using PyArrowFile"""
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        file_location = os.path.join(tmpdirname, "foo.txt")
+        with open(file_location, "wb") as f:
+            f.write(b"foo")
+
+        # Confirm that the file initially exists
+        assert os.path.exists(file_location)
+
+        # Instantiate the input file
+        absolute_file_location = os.path.abspath(file_location)
+        input_file = PyArrowFile(location=f"{absolute_file_location}")
+
+        # Test opening and reading the file
+        f = input_file.open()
+        data = f.read()
+        assert data == b"foo"
+        assert len(input_file) == 3
+
+
+def test_pyarrow_output_file():
+    """Test writing a file using PyArrowFile"""
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        file_location = os.path.join(tmpdirname, "foo.txt")
+
+        # Instantiate the output file
+        absolute_file_location = os.path.abspath(file_location)
+        output_file = PyArrowFile(location=f"{absolute_file_location}")
+
+        # Create the output file and write to it
+        f = output_file.create()
+        f.write(b"foo")
+
+        # Confirm that bytes were written
+        with open(file_location, "rb") as f:
+            assert f.read() == b"foo"
+
+        assert len(output_file) == 3
+
+
+def test_pyarrow_invalid_scheme():
+    """Test that a ValueError is raised if a location is provided with an invalid scheme"""
+
+    with pytest.raises(ValueError) as exc_info:
+        PyArrowFile("foo://bar/baz.txt")
+
+    assert ("Unrecognized filesystem type in URI") in str(exc_info.value)
+
+    with pytest.raises(ValueError) as exc_info:
+        PyArrowFile("foo://bar/baz.txt")
+
+    assert ("Unrecognized filesystem type in URI") in str(exc_info.value)
+
+
+def test_pyarrow_violating_input_stream_protocol():
+    """Test that a TypeError is raised if an input file is provided that violates the InputStream protocol"""
+
+    # Missing seek, tell, closed, and close
+    input_file_mock = MagicMock(spec=["read"])
+
+    # Create a mocked filesystem that returns input_file_mock
+    filesystem_mock = MagicMock()
+    filesystem_mock.open_input_file.return_value = input_file_mock
+
+    input_file = PyArrowFile("foo.txt")
+    input_file._filesystem = filesystem_mock
+    with pytest.raises(TypeError) as exc_info:
+        input_file.open()
+
+    assert ("Object of type") in str(exc_info.value)
+    assert ("returned from PyArrowFile.open does not match the InputStream protocol.") in str(exc_info.value)
+
+
+def test_pyarrow_violating_output_stream_protocol():
```

Comment:
it would be nice to add a test with something like minio to ensure we are getting the semantics we want on metastores.

Reply:
Do you mean like adding a test suite that launches a local minio server and uses it to mock s3?

Reply:
Thats what I was thinking.  The comment you mentioned above about having to adjust constructor above really covers my main concern in terms of consistent user experience.  It might be too heavy weight here.  I think for different implementations that access s3 (and other remote file systems) we might just need to accept that there might be quirks in behavior.